nhs/data/join_on_sa1.py

```diff
@@ -0,0 +1,219 @@
+    logger.info(f"Configuration loaded: {config}")
+    return config
+
+
+def read_shapefile(path: str) -> int:
+    """Read shapefile and return total number of rows."""
+    total_rows = len(pl.read_parquet(path))
+    logger.info(f"Total number of rows in the shapefile: {total_rows}")
+    return total_rows
+
+
+def read_csv(path: str) -> pl.DataFrame:
+    """Read CSV file and return as Polars DataFrame."""
+    df = pl.read_csv(path)
+    logger.info(df.head())
+    return df
+
+
+def prepare_points(df: pl.DataFrame, crs: str) -> Any:
+    """Prepare points data for geospatial analysis."""
+    if "LONGITUDE" in df.columns and "LATITUDE" in df.columns:
+        df = df.rename({"LONGITUDE": "x", "LATITUDE": "y"})
+
+    gdf = gpd.GeoDataFrame(
+        df.to_pandas(),
+        geometry=gpd.points_from_xy(df["x"], df["y"]),
+        crs="EPSG:7844",  # Assuming input is in WGS84
+    )  # type: ignore
+    projected_crs = CRS.from_string(crs)
+    return gdf.to_crs(projected_crs)
+
+
+def process_chunk(map_data: Any, chunk: Any) -> pl.DataFrame:
+    """Process a chunk of point data, joining with area data."""
+    pnts_with_area = gpd.sjoin(chunk, map_data, how="left", predicate="within")
+    missing_points = pnts_with_area[pnts_with_area["SA2_NAME21"].isnull()]
+
+    if not missing_points.empty:
+        nearest_join = gpd.sjoin_nearest(
+            missing_points[["geometry"]], map_data, how="left"
+        )
+        pnts_with_area.loc[missing_points.index, "SA2_NAME21"] = nearest_join[
+            "SA2_NAME21"
+        ]
+        pnts_with_area.loc[missing_points.index, "SA1_CODE21"] = nearest_join[
+            "SA1_CODE21"
+        ]
+
+    result = pnts_with_area[["x", "y", "SA2_NAME21", "SA1_CODE21"]]
+    result = result.rename(columns={"SA2_NAME21": "area", "SA1_CODE21": "area_code"})
+    return pl.DataFrame(
+        result.to_dict()
+    )  # Convert to dict first to ensure compatibility with Polars
+
+
+def parallel_process(pnts_gdf: Any, map_data: Any, num_cores: int) -> pl.DataFrame:
+    """Parallel process point data."""
+    chunks = np.array_split(pnts_gdf, num_cores)
+    with Pool(num_cores) as pool:
+        all_results = pool.map(lambda chunk: process_chunk(map_data, chunk), chunks)
+    return pl.concat(all_results)
+
+
+def random_distribution(df: pl.DataFrame, num_iterations: int = 10) -> pl.DataFrame:
+    """Randomly shuffle lat-long pairs within their respective areas."""
+    model_df = df.clone()
+    model_df = model_df.with_columns(
+        [pl.col("y").alias("original_lat"), pl.col("x").alias("original_lon")]
+    )
+
+    for _ in range(num_iterations):
+        # Loop over unique area_code values and shuffle lat-long pairs within each group
+        area_codes = model_df["area_code"].unique().to_list()
+        for area_code in area_codes:
+            area_df = model_df.filter(pl.col("area_code") == area_code)
+            shuffled_pairs = area_df[["x", "y"]].sample(frac=1).to_dict(as_series=False)
```

Comment:
What does .sample(frac=1) do, is it just a normal shuffle of all the pairs within the area? And if so I think we should seed it.

Reply:
Yeah that's right -- I've set a seed param to be 42.